exec: implement the case operator

jordanlewis opened the pull request:
This PR adds support for projections of CASE expressions in the vectorized engine. It only adds support for the form `CASE WHEN x THEN y ... [ELSE z]`, not the `CASE foo WHEN x THEN y ...` form.

Release note: None

jordanlewis committed:
exec: implement the case operator

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38999)
<!-- Reviewable:end -->

rohany reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/exec/bool_vec_to_sel.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAdbOC41Gc-Ll7XFrD:-LkAdbOC41Gc-Ll7XFrE:b-ma9eis) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/bool_vec_to_sel.go#L69)):*
> ```Go
> }
> 
> func boolVecToSelWithOldSel(col []bool, oldSel, newSel []uint16) uint16 {
> ```

since these are only used here, no need to pull them out.
___
*[pkg/sql/exec/buffer.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAdunz45w18rJYJD6E:-LkAdunz45w18rJYJD6F:b-7ymzek) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/buffer.go#L21)):*
> ```Go
> // bufferOp is an operator that buffers a single batch at a time from an input,
> // and makes it available to be read multiple times by downstream consumers.
> type bufferOp struct {
> ```

This seems interesting -- I could probably look into using this for the builtins.
___
*[pkg/sql/exec/buffer.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAQkTx-zsYye2eTVZo:-LkAQkTx-zsYye2eTVZp:b8sjtj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/buffer.go#L36)):*
> ```Go
> 		input: input,
> 		batch: &selectionBatch{
> 			sel: make([]uint16, coldata.BatchSize),
> ```

probably account for this memory used as a StaticMemoryOperator
___
*[pkg/sql/exec/const_tmpl.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAeZo-DWnoVO3lWjhp:-LkAeZo-DWnoVO3lWjhq:bt9xjwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/const_tmpl.go#L81)):*
> ```Go
> }
> 
> func (c const_TYPEOp) Next(ctx context.Context) coldata.Batch {
> ```

For clarity, its probably best to move these auxiliary changes out of this PR.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/exec/bool_vec_to_sel.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAdbOC41Gc-Ll7XFrD:-LlHvfeaCuWcW_TFCTok:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/bool_vec_to_sel.go#L69)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

since these are only used here, no need to pull them out.
</blockquote></details>

Done.
___
*[pkg/sql/exec/buffer.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAQkTx-zsYye2eTVZo:-LlI01StBBNrwJujpRF6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/buffer.go#L36)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

probably account for this memory used as a StaticMemoryOperator
</blockquote></details>

Done.
___
*[pkg/sql/exec/const_tmpl.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LkAeZo-DWnoVO3lWjhp:-LlHvvoK8vORrI4mVy_d:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cd60164d96220f62e2cbd09463cd49124fb173/pkg/sql/exec/const_tmpl.go#L81)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

For clarity, its probably best to move these auxiliary changes out of this PR.
</blockquote></details>

Good call, done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice work! A bunch of nits, but it <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 7 files at r1, 1 of 7 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-:-LlJTCjrEQd3HfgajWln:b7222l4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 790 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJMHpOBq1RS_b54355:-LlJMHpOBq1RS_b54356:b-bnalw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L790)):*
> ```Go
> 		caseOps := make([]exec.Operator, len(t.Whens))
> 		for i, when := range t.Whens {
> 			// Write all cond bools to a single column. The operators will
> ```

[nit]: could you please expand the first sentence?
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 796 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJMa3s3sakbcfj6BA1:-LlJMa3s3sakbcfj6BA2:bhaepg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L796)):*
> ```Go
> 			// the WHEN or THEN expressions on tuples that have already matched an
> 			// arm.
> 			var whenUsed, thenUsed int
> ```

[nit]: I'd call these `whenMemUsed` and `thenMemUsed` because initially I thought they were a boolean whether "when was used."
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJOJSf28aehN7u4i4L:-LlJOJSf28aehN7u4i4M:b-wpill3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L797)):*
> ```Go
> 			// arm.
> 			var whenUsed, thenUsed int
> 			caseOps[i], resultIdx, ct, whenUsed, err = planProjectionOperators(ctx, when.Cond.(tree.TypedExpr), columnTypes,
> ```

[nit]: I think it would be better to explicitly assign to `ct` as with types of expressions. Also, it might be worth it to have a comment why we're slicing `ct` in certain way (like `ct[:len(ct)-1]` below).
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 798 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJN22a6KQpUulWQ5cA:-LlJN22a6KQpUulWQ5cB:b-yo273u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L798)):*
> ```Go
> 			var whenUsed, thenUsed int
> 			caseOps[i], resultIdx, ct, whenUsed, err = planProjectionOperators(ctx, when.Cond.(tree.TypedExpr), columnTypes,
> 				buffer)
> ```

[nit]: it is slightly odd to see a single argument on the new line.
___
*[pkg/sql/exec/buffer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQVib7rAn6CLjhPTn:-LlJQVib7rAn6CLjhPTo:b-qlev9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L42)):*
> ```Go
> }
> 
> func (b bufferOp) EstimateStaticMemoryUsage() int {
> ```

[nit]: some methods have pointer receivers, some have value receivers.
___
*[pkg/sql/exec/buffer.go, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJPOo94cnxHXtyKkvc:-LlJPOo94cnxHXtyKkvd:bd63hzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L60)):*
> ```Go
> }
> 
> // advance reads the next input from the buffers input into its buffer, and
> ```

[nit]: I'd adjust the comment a bit, maybe like "advance reads the next batch from the input into the buffer ..."
___
*[pkg/sql/exec/case.go, line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQe0I4uTLKbzU1oqD:-LlJQe0I4uTLKbzU1oqE:ba6vora) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L20)):*
> ```Go
> )
> 
> type caseOp struct {
> ```

Should `caseOp` estimate its memory usage?
___
*[pkg/sql/exec/case.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQFsxDOw-ttCJ5w7F:-LlJQFsxDOw-ttCJ5w7G:bjihfo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L30)):*
> ```Go
> 	typ     types.T
> 
> 	origSel []uint16
> ```

[nit]: I'd probably add a short comment here why it's needed.
___
*[pkg/sql/exec/case.go, line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJSx0G-8UKVy9Vtc7e:-LlJSx0G-8UKVy9Vtc7f:bq1dhpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L34)):*
> ```Go
> 
> // NewCaseOp returns an operator that runs a case statement.
> // buffer is a bufferOp that will return the input batch repeatedly.
> ```

[nit]: I'd highlight the argument names by indenting or putting a dash in front of each since there is a longer explanation for `caseOps`.
___
*[pkg/sql/exec/case.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJR7T3AGgvHYrVvqj7:-LlJR7T3AGgvHYrVvqj8:bfuha0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L82)):*
> ```Go
> 		// already matched.
> 		batch := c.caseOps[i].Next(ctx)
> 		// 1 2 3 4
> ```

[nit]: it feels like this example should be below the explanation. Also, it would be nice to indicate each "level" with the corresponding variable name used below (i.e. the first level is `oldSel` before the update, etc).
___
*[pkg/sql/logictest/testdata/logic_test/conditional, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJSek3AomXNNFotXO9:-LlJSek3AomXNNFotXOA:bnxlulg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/logictest/testdata/logic_test/conditional#L1)):*
> ```
> # LogicTest: local local-opt fakedist fakedist-metadata fakedist-opt local-vec
> ```

Might as well add `fakedist-vec` config since it's been merged.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 790 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJMHpOBq1RS_b54355:-LlNp3687WXOakBPijQo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L790)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: could you please expand the first sentence?
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 796 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJMa3s3sakbcfj6BA1:-LlNp2vj-1saU3QQ9EzF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L796)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I'd call these `whenMemUsed` and `thenMemUsed` because initially I thought they were a boolean whether "when was used."
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJOJSf28aehN7u4i4L:-LlNp3_yCSWlnI4g4OcZ:bu45mqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L797)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think it would be better to explicitly assign to `ct` as with types of expressions. Also, it might be worth it to have a comment why we're slicing `ct` in certain way (like `ct[:len(ct)-1]` below).
</blockquote></details>

Thanks, thinking about this actually made me realize that there is a pretty tricky bug here that we're not prepared to fix yet. I was assuming that the result of the WHEN expression would always go into the next slot, and the result of THEN would go into the following slot.

But since we just use the normal plan projection functions, if a WHEN or THEN has a chain of projections required, it will be kind of arbitrary where the result goes, and we'll have no real chance of making sure the output always lines up.

I think to fix this we will need to refactor planning significantly. I think projection planning operators will need to take in the desired output slot as input... and we might need to keep a global map around of what slots have been used and so on. :/
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 798 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJN22a6KQpUulWQ5cA:-LlNp2hFEhaEBxT50ODV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L798)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it is slightly odd to see a single argument on the new line.
</blockquote></details>

Done.
___
*[pkg/sql/exec/buffer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQVib7rAn6CLjhPTn:-LlNossE4xUStKjqBH5_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L42)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: some methods have pointer receivers, some have value receivers.
</blockquote></details>

Done.
___
*[pkg/sql/exec/buffer.go, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJPOo94cnxHXtyKkvc:-LlNoyJQ43bUBVJgSkcP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L60)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I'd adjust the comment a bit, maybe like "advance reads the next batch from the input into the buffer ..."
</blockquote></details>

Done.
___
*[pkg/sql/exec/case.go, line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQe0I4uTLKbzU1oqD:-LlNnkK72NBy051LFUJU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L20)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Should `caseOp` estimate its memory usage?
</blockquote></details>

Done.
___
*[pkg/sql/exec/case.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQFsxDOw-ttCJ5w7F:-LlNnkfA7l3tyDGBxA0w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L30)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I'd probably add a short comment here why it's needed.
</blockquote></details>

Done.
___
*[pkg/sql/exec/case.go, line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJSx0G-8UKVy9Vtc7e:-LlNnl2UEr98TW6JSIkT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L34)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I'd highlight the argument names by indenting or putting a dash in front of each since there is a longer explanation for `caseOps`.
</blockquote></details>

Done.
___
*[pkg/sql/exec/case.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJR7T3AGgvHYrVvqj7:-LlNnmZm7w91a_rFSNIG:b-x0mfiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L82)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it feels like this example should be below the explanation. Also, it would be nice to indicate each "level" with the corresponding variable name used below (i.e. the first level is `oldSel` before the update, etc).
</blockquote></details>

Yeah, this was not fleshed out. I think I used it as a diagram for myself during implementation :) Thank you for pointing it out. I improved the example.
___
*[pkg/sql/logictest/testdata/logic_test/conditional, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJSek3AomXNNFotXO9:-LlNq0EoADFmPSnkGW6P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/logictest/testdata/logic_test/conditional#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Might as well add `fakedist-vec` config since it's been merged.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJOJSf28aehN7u4i4L:-LlNsB6R7ZPMjFbCK28d:b-42qb52) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L797)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Thanks, thinking about this actually made me realize that there is a pretty tricky bug here that we're not prepared to fix yet. I was assuming that the result of the WHEN expression would always go into the next slot, and the result of THEN would go into the following slot.

But since we just use the normal plan projection functions, if a WHEN or THEN has a chain of projections required, it will be kind of arbitrary where the result goes, and we'll have no real chance of making sure the output always lines up.

I think to fix this we will need to refactor planning significantly. I think projection planning operators will need to take in the desired output slot as input... and we might need to keep a global map around of what slots have been used and so on. :/
</blockquote></details>

I see. Please leave a TODO here about this case.
___
*[pkg/sql/exec/buffer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQVib7rAn6CLjhPTn:-LlNsd1h1VELFSnWz2rO:b-4or2of) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L42)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`EstimateStatisMemoryUsage` and `Init` still have value receivers.
___
*[pkg/sql/exec/case.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJR7T3AGgvHYrVvqj7:-LlNt49_8fW9uD7QrBih:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/case.go#L82)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yeah, this was not fleshed out. I think I used it as a diagram for myself during implementation :) Thank you for pointing it out. I improved the example.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJOJSf28aehN7u4i4L:-LlOZV5t1-Bv5VHwiapg:b-rzzgx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L797)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I see. Please leave a TODO here about this case.
</blockquote></details>

Well, I can't merge this until it is fixed.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
It's RFAL again, and based off of #39792 now.

The big change is that it no longer tries to do any fancy thing to ensure that all of the THEN arms get projected to the same column. Instead, it projects each THEN arm to its own column, and has a copy phase of moving each of the matched tuples from each THEN arm to the final output column.

yuzefovich reviewed (approved):
Looked only at the case operator commit, and I like the new approach. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 9 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-:-LmuM8Zj4Qj1NnvMOU_1:bozoreq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/col/coldata/vec_tmpl.go, line 108 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuIWWTDBp3HlCDOHgs:-LmuIWWTDBp3HlCDOHgt:b-ei1ljl) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/col/coldata/vec_tmpl.go#L108)):*
> ```Go
> 	toColSliced := execgen.SLICE(toCol, int(args.DestIdx), n)
> 	for i := range sel[args.SrcStartIdx:args.SrcEndIdx] {
> 		selIdx := sel[int(args.SrcStartIdx)+i]
> ```

[nit]: why moving this onto a separate line? Is it faster this way?
___
*[pkg/sql/exec/buffer.go, line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuJpEW6wuwE4Ao20ov:-LmuJpEW6wuwE4Ao20ow:b-4923yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/buffer.go#L22)):*
> ```Go
> // bufferOp is an operator that buffers a single batch at a time from an input,
> // and makes it available to be read multiple times by downstream consumers.
> type bufferOp struct {
> ```

[nit]: we should add checks that `bufferOp` is a `StaticMemoryOperator` and that `selectionBatch` is `coldata.Batch`.
___
*[pkg/sql/exec/case.go, line 20 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuKOeG9sKGPTak5maR:-LmuKOeG9sKGPTak5maS:b-41vwxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/case.go#L20)):*
> ```Go
> )
> 
> type caseOp struct {
> ```

ditto for the check.
___
*[pkg/sql/exec/case.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuKUOy0OiI2EJT6RCd:-LmuKUOy0OiI2EJT6RCe:b-fomv5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/case.go#L41)):*
> ```Go
> 
> func (c *caseOp) Child(nth int) OpNode {
> 	return c.buffer
> ```

[nit]: should we check value of `nth`?
___
*[pkg/sql/exec/case.go, line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuL5Mz7R-rqc2kCbDD:-LmuL5Mz7R-rqc2kCbDE:b-l7co2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/case.go#L133)):*
> ```Go
> 		inputCol := c.buffer.batch.Batch.ColVec(c.thenIdxs[i])
> 		// Copy the results into the output vector,
> 		outputCol.Copy(coldata.CopyArgs{
> ```

[nit]: `s/,/./` (above).
___
*[pkg/sql/exec/simple_project.go, line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuLt9r4d5j85xo_g8G:-LmuLt9r4d5j85xo_g8H:b-qzjclf) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/simple_project.go#L52)):*
> ```Go
> func (b *projectingBatch) ColVecs() []coldata.Vec {
> 	execerror.VectorizedInternalPanic("projectingBatch doesn't support ColVecs()")
> 	panic("unreachable")
> ```

This should fail the linter.
___
*[pkg/sql/logictest/testdata/logic_test/conditional, line 1 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuM0Y25kRmdDntxl2L:-LmuM0Y25kRmdDntxl2M:bl43itx) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/logictest/testdata/logic_test/conditional#L1)):*
> ```
> # LogicTest: local fakedist fakedist-metadata local-vec fakedist-vec
> ```

[super nit]: usually `local`s go before `fakedist`s.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/col/coldata/vec.go, line 59 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmzjXzY5OVUlbkD3p7s:-LmzjXzY5OVUlbkD3p7t:b-updxjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a646f1ef008ca6a52cae7673edc33e31b4a47f08/pkg/col/coldata/vec.go#L59)):*
> ```Go
> 	// selection vector is applied to the source vector, but the results are
> 	// copied densely into the destination vector.
> 	SelOnDest bool
> ```

cc @asubiotto - could you take a look at this as it changes the `CopyArgs` struct, making it slightly more confusing.
___
*[pkg/col/coldata/vec_tmpl.go, line 108 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuIWWTDBp3HlCDOHgs:-Lmzh-VI0LhcbUU1bbRy:b5h4bwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/col/coldata/vec_tmpl.go#L108)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: why moving this onto a separate line? Is it faster this way?
</blockquote></details>

No idea why I did this, reverted.
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJOJSf28aehN7u4i4L:-LmzjU-TEUWiJ50MLp6_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/distsqlrun/column_exec_setup.go#L797)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Well, I can't merge this until it is fixed.
</blockquote></details>

Done.
___
*[pkg/sql/exec/buffer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQVib7rAn6CLjhPTn:-LmzhCtrA6LIAjO8hwqD:bpurhpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L42)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

`EstimateStatisMemoryUsage` and `Init` still have value receivers.
</blockquote></details>

Oops, thanks. Should be fixed now.
___
*[pkg/sql/exec/buffer.go, line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuJpEW6wuwE4Ao20ov:-LmzhY3B-Nu57u8A7U4W:b-i43c46) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/buffer.go#L22)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: we should add checks that `bufferOp` is a `StaticMemoryOperator` and that `selectionBatch` is `coldata.Batch`.
</blockquote></details>

Done, and added the missing comment to `selectionBatch` while I was at it.
___
*[pkg/sql/exec/case.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuKUOy0OiI2EJT6RCd:-Lmzj7pz-7hC_CSVtExh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/case.go#L41)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: should we check value of `nth`?
</blockquote></details>

Done.
___
*[pkg/sql/exec/case.go, line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuL5Mz7R-rqc2kCbDD:-LmzjHnc7pMRo5y350IQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/case.go#L133)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/,/./` (above).
</blockquote></details>

Done.
___
*[pkg/sql/exec/simple_project.go, line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuLt9r4d5j85xo_g8G:-LmzjN1_2y-RqbNrvH-f:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/simple_project.go#L52)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This should fail the linter.
</blockquote></details>

Reverted.
___
*[pkg/sql/logictest/testdata/logic_test/conditional, line 1 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuM0Y25kRmdDntxl2L:-LmzjQfa3h78VURh1hMe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/logictest/testdata/logic_test/conditional#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[super nit]: usually `local`s go before `fakedist`s.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/col/coldata/vec_tmpl.go, line 108 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuIWWTDBp3HlCDOHgs:-LmzkQz99wl8D-2a-dHn:bwlzmft) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/col/coldata/vec_tmpl.go#L108)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

No idea why I did this, reverted.
</blockquote></details>

Oh, upon reverting, it doesn't compile. It's a dumb reason - the non `SelOnDest` case only uses `i`, and the `SelOnDest` version only uses `selIdx`, so you can't get them to both compile at the same time without doing what I did above.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/col/coldata/vec.go, line 59 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmzjXzY5OVUlbkD3p7s:-Lmzm7Qp4wV5U9ikygdC:bapkimo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a646f1ef008ca6a52cae7673edc33e31b4a47f08/pkg/col/coldata/vec.go#L59)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

cc @asubiotto - could you take a look at this as it changes the `CopyArgs` struct, making it slightly more confusing.
</blockquote></details>

:+1: I'm generally against adding more complexity to the `CopyArgs` struct if we can avoid it but I guess the alternative here is to copy code, so this <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 9 files at r9, 1 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/exec/buffer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LlJQVib7rAn6CLjhPTn:-LmztpKp5biAoleW7G-M:b-61oeuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f6186833bfb16c47ea9036dd88dab99f1972c9/pkg/sql/exec/buffer.go#L42)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Oops, thanks. Should be fixed now.
</blockquote></details>

I still see value receivers on those two methods :) (maybe you haven't pushed the latest update yet?)
___
*[pkg/sql/exec/buffer.go, line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-LmuJpEW6wuwE4Ao20ov:-Lmztd-59xwJ0cABMPUT:bu2pckm) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f5645f9a291ff7c43e7996f5fd4203b233d257/pkg/sql/exec/buffer.go#L22)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Done, and added the missing comment to `selectionBatch` while I was at it.
</blockquote></details>

What I meant was adding
```
var _ StaticMemoryOperator = &bufferOp{}
```


<!-- Sent from Reviewable.io -->

jordanlewis commented:
I haven't pushed as there is a new bug I'm trying to understand. I'll push for now though.

jordanlewis committed:
exec: ops extend input widths even w/ empty input

Previously, operators that needed extra space in their input batches
would only extend them if they received more than 0 input tuples. This
was problematic in some cases when operator chains would be reused. Now,
we always extend input batch widths if necessary.

Note that this solution does not feel satisfying as it's brittle. We'll
hopefully figure out a more sane way to do this across the board soon.

Release note: None

jordanlewis reviewed (commented):
RFAL again - I added a first commit to fix the issue I mentioned in slack, that operators didn't consistently extend their input batches if they had 0 input tuples.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999#-:-Ln-exVBARjZHYYaRTzR:b-c6fr6x)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 16 of 16 files at r11, 8 of 8 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38999)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs all

bors r+
